components/scream/src/physics/p3/p3_functions_upwind_impl.hpp

```diff
@@ -45,8 +46,8 @@ ::calc_first_order_upwind_step (
     Kokkos::PerTeam(team), [&] () {
       const Int k = k_top_pack;
       if (nk % Spack::n != 0) {
-        const auto mask =
-          scream::pack::range<IntSmallPack>(k_top_pack*Spack::n) >= nk;
+        auto range_pack = scream::pack::range<IntSmallPack>(k_top_pack*Spack::n);
```

Comment:
@ambrad , I found that I needed to make this minor tweak in order to handle the case where kmin begins mid-pack.

Reply:
First, this mostly makes sense; see next paragraph. There's a complicated reason why the second condition is not needed in the micro-app, despite the condition holding in its tests. (Roughly, it has to do with the state of V, r, rho on valid but unused scalar indices vs their state on invalid scalar indices, where by invalid I mean `>= nk`.) However, it's much better to include the second condition to harden this function as a true library routine.

However, I believe the second condition is not quite correct in the context of the `nk % Spack::n != 0` condition; it must be checked regardless of `nk % Spack::n`. I believe that with the second condition included, we need to execute the whole block always:
```c++
{ // <-- mod condition removed
  const auto range_pack = scream::pack::range<IntSmallPack>(k_top_pack*Spack::n); // <-- don't forget `const` :)
  const auto mask = range_pack < kmin_scalar || range_pack >= nk;
  if (mask.any())  // <-- debatable whether this is a micro-opt or not
    for (int f = 0; f < nfield; ++f)
      (*flux[f])(k_top_pack).set(mask, 0);
}
```
You must have written a nice unit test to catch this. Either I'm wrong about removing the mod condition, or the unit test can be strengthened one notch more.

Nice work!

Reply:
@ambrad , thanks for taking a close look at this.

You're exactly right. Even if ```nk % Spack::n == 0``` that doesn't necessarily mean that the active k-range falls exactly along pack boundaries. We could have 8 values at the end the top pack and 8 values at the beginning of the next pack. If pack::n is 16, then this code would fail.